oldschnorr.go

```diff
@@ -0,0 +1,149 @@
+type SchnorrSignature struct {
+	signature [64]byte
+}
+
+// Serialize returns a 64 byte serialized signature
+func (signature *SchnorrSignature) Serialize() [64]byte {
+	return signature.signature
+}
+
+// DeserializeSchnorrSignature deserializes a 64 byte serialized schnorr signature into a SchnorrSignature type.
+func DeserializeSchnorrSignature(serializedSignature [64]byte) SchnorrSignature {
+	signature := SchnorrSignature{}
+	signature.signature = serializedSignature
+	return signature
+}
+
+// SchnorrPublicKey generates a PublicKey for the corresponding private key.
+func (key *PrivateKey) SchnorrPublicKey() (*SchnorrPublicKey, error) {
+	pubkey := SchnorrPublicKey{}
+	cPtrPrivateKey := (*C.uchar)(&key.privateKey[0])
+	ret := C.secp256k1_ec_pubkey_create(context, &pubkey.pubkey, cPtrPrivateKey)
+	if ret != 1 {
+		return nil, errors.New("failed Generating an SchnorrPublicKey. You should call `DeserializePrivateKey` before calling this")
+	}
+	return &pubkey, nil
+}
+
+// SchnorrSign creates a schnorr signature using the private key and the input hashed message.
+// Notice: the [32] byte array *MUST* be a hash of a message.
+func (key *PrivateKey) SchnorrSign(hash *Hash) (*SchnorrSignature, error) {
+	signature := SchnorrSignature{}
+	cPtrSig := (*C.uchar)(&signature.signature[0])
+	cPtrHash := (*C.uchar)(&hash[0])
+	cPtrPrivKey := (*C.uchar)(&key.privateKey[0])
+	ret := C.secp256k1_schnorr_sign(context, cPtrSig, cPtrHash, cPtrPrivKey, nil, nil)
+	if ret != 1 {
+		return nil, errors.New("failed Signing. You should call `DeserializePrivateKey` before calling this")
+	}
+	return &signature, nil
+}
+
+// SchnorrVerify verifies a schnorr signature using the public key and the input hashed message.
+// Notice: the [32] byte array *MUST* be a hash of a message you hashed yourself.
+func (key *SchnorrPublicKey) SchnorrVerify(hash *Hash, signature SchnorrSignature) bool {
+	cPtrHash := (*C.uchar)(&hash[0])
+	cPtrSig := (*C.uchar)(&signature.signature[0])
+	return C.secp256k1_schnorr_verify(context, cPtrSig, cPtrHash, &key.pubkey) == 1
+}
+
+// DeserializeSchnorrPubKey deserializes a serialized schnorr public key, verifying it's valid.
+// it supports both compressed(33 bytes) and uncompressed(65 bytes) public keys.
+// it does not support hybrid(65 bytes) keys.
+func DeserializeSchnorrPubKey(serializedPubKey []byte) (*SchnorrPublicKey, error) {
+	key := SchnorrPublicKey{}
+	cPtr := (*C.uchar)(&serializedPubKey[0])
+	cLen := C.size_t(len(serializedPubKey))
+	if !supportedKey(serializedPubKey) {
+		return nil, errors.New("unsupported public key format")
+	}
+	ret := C.secp256k1_ec_pubkey_parse(C.secp256k1_context_no_precomp, &key.pubkey, cPtr, cLen)
+	if ret != 1 {
+		return nil, errors.New("failed parsing the public key")
+	}
+	return &key, nil
+}
+
+// SerializeCompressed serializes a schnorr public key into a compressed form (33 bytes)
+func (key *SchnorrPublicKey) SerializeCompressed() (compressedPubKey [33]byte, err error) {
```

Comment:
Should I really create a type for serialized compressed public key? and uncompressed public key? what about a serialized Signature?
would you prefer maybe if it will return a slice instead of an array?

Reply:
Any thing that is a "Type" in the real-world should have a type in code.

Reply:
Done